[Feature] Add a section list and changePassword Modal to Setting screen

JongtaekChoi opened the pull request:
## Description

I added the section list and section item interface to the Setting screen. And I implement the changing password modal. 

## Related Issues

#29 

## Tests

I added the following tests:

- render Setting screen 
- press login state  section item 
- render ChangePwScreen
- press change password button
- text change of password text inputs 

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk-mobile/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dooboolab/hackatalk-mobile?pullRequest=58) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dooboolab/hackatalk-mobile?pullRequest=58) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dooboolab/hackatalk-mobile?pullRequest=58) it.</sub>

Sandwichj commented on src/components/screen/Setting/ChangePwView.tsx:
```diff
@@ -0,0 +1,96 @@
+  const [newPw, setNewPw] = useState('');
+  const [newPwCheck, setNewPwCheck] = useState('');
+  const checkCurrent = (): void => {
+    checkCurrentPw().then((verified) => {
```
The naming `check` means A, B or more.
And this function has two roles `Updating state` and `Validating password`.
So, if the result value is `boolean`, the roles should be separated for readability and intuition.

For example,
```js
isValid~()  // Validate password
updated~Verified(isValid~()) // Update state
```

It makes code clear. (ref. [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle))

JongtaekChoi commented:
> Try to fill out the description on your `PR`

I did it!

hyochan commented on src/components/screen/ProfileUpdate.tsx:
```diff
@@ -79,6 +79,12 @@ function Screen(props: Props): React.ReactElement {
   };
 
+  const onSetting = (): void => {
+    if (props.navigation) {
```
It'd be better to use `const navigation = useNavigation();`.

JongtaekChoi replied:
Actually it's now unnecessary. I will remove it. Thanks.

hyochan commented on src/components/screen/Setting/ChangePwView.tsx:
```diff
@@ -0,0 +1,96 @@
+
+export interface Props {
+    theme: DefaultTheme;
+    close(): void;
```
I don't think this is a good idea if this is a `screen` component. It'd be better to pass the param to `react-navigation` and get it from the `route` param.

hyochan commented on src/components/screen/Setting/ChangePwView.tsx:
```diff
@@ -0,0 +1,96 @@
+export interface Props {
+    theme: DefaultTheme;
+    close(): void;
+    validateCurrentPw(text: string): Promise<boolean>;
```
same here.

hyochan replied:
What screen prop has is simply below.
```
interface Props extends ThemeProps<DefaultTheme> {
  navigation: DefaultNavigationProps;
}
```

I currently don't think there could be more than this.

hyochan commented on src/components/screen/Setting/ChangePwView.tsx:
```diff
@@ -0,0 +1,96 @@
+import { getString } from '../../../../STRINGS';
+
+export interface Props {
+    theme: DefaultTheme;
```
This is currently deprecated. We are using `const theme = useThemeContext()`.

JongtaekChoi replied:
I'd remove it.

hyochan commented on src/components/screen/Setting/index.tsx:
```diff
@@ -0,0 +1,87 @@
+}
+
+interface Props {
+  navigation: DefaultNavigationProps<'Setting'>;
```
Good one here. You are using it correctly.

JongtaekChoi replied:
Thanks. But, is it needed? In this https://github.com/dooboolab/hackatalk-mobile/pull/58#discussion_r347398884 you said that it'd be better to use `const navigation = useNavigation();`

hyochan replied:
This is needed if you want to strictly check the state param passed down to the `Setting` screen.

Oh yes, and sorry for confusing you. I wanted to tell you that `useNavigation` would be efficient in such cases if you need the navigation props in `shared` components. Then you won't have to pass down the `navigation` props every time. In this case, `useNavigation` is not a good idea.

hyochan commented on src/components/screen/Setting/index.tsx:
```diff
@@ -0,0 +1,87 @@
+
+interface Props {
+  navigation: DefaultNavigationProps<'Setting'>;
+  theme: DefaultTheme;
```
As said above, this can be removed.

JongtaekChoi replied:
I'd remove it. Thanks.

hyochan commented on src/components/screen/Setting/index.tsx:
```diff
@@ -0,0 +1,87 @@
+}
+
+function SettingScreen(props: Props): React.ReactElement {
+  const { theme } = props;
```
Replace this with `const { theme } = useThemeContext();`

JongtaekChoi replied:
I will change it. Thx.

hyochan commented on src/components/screen/Setting/index.tsx:
```diff
@@ -0,0 +1,87 @@
+
+  const settings: SectionListData<SettingsOption>[] = [
+    {
+      title: '계정정보',
```
Try to localize this with `getString` since this looks constant.

hyochan commented on src/components/screen/Setting/index.tsx:
```diff
@@ -0,0 +1,87 @@
+      title: '계정정보',
+      data: [{
+        icon: IC_FACEBOOK,
+        label: 'facebook으로 로그인 됨',
```
Same here! Localize~

JongtaekChoi commented:
@hyochan I added some changes.
1. ChangePwView was moved to the shared folder and renamed to ChangePwModal
2. Do I need to remove the navigation options of the 'ProfileUPdate' stack screen?

hyochan commented on src/components/screen/Setting/index.tsx:
```diff
@@ -0,0 +1,100 @@
+  );
+}
+
+export default withTheme(SettingScreen);
```
You don't need `withTheme` with hook.

hyochan commented on src/components/screen/Setting/styles.ts:
```diff
@@ -0,0 +1,72 @@
+export const ItemIcon = styled.Image`
+width: 32px;
+height: 32px;
+resize-mode: contain;
```
This isn't a style!

hyochan commented on src/components/shared/ChangePwModal.tsx:
```diff
@@ -0,0 +1,133 @@
+  );
+});
+
+export default withTheme(ModalWrapper);
```
No need to wrap with `withTheme` in hook!

JongtaekChoi commented:
@hyochan Thanks for your work. I have checked the problem of 'blinking'. It occurred when closing the ChangePwModal. I'll try to fix this.

hyochan commented:
@JongtaekChoi How is this going?

JongtaekChoi commented:
I think it caused by SafeAreaView.
When I changed SafeAreaView to normal View it doesn't occur.

![image](https://user-images.githubusercontent.com/17980230/70403623-bbf5c580-1a7a-11ea-9657-9f95c76ae03f.png)
[refer to Reat Navigation Docs](https://reactnavigation.org/docs/en/next/handling-safe-area.html)

Fortunately, [expo support `react-native-safe-area-context` library.](https://docs.expo.io/versions/v35.0.0/sdk/safe-area-context/)

I will try this method within today and add commit if it works.

> @JongtaekChoi How is this going?

JongtaekChoi commented:
> I think it caused by SafeAreaView.
> When I changed SafeAreaView to normal View it doesn't occur.
> 
> ![image](https://user-images.githubusercontent.com/17980230/70403623-bbf5c580-1a7a-11ea-9657-9f95c76ae03f.png)
> [refer to Reat Navigation Docs](https://reactnavigation.org/docs/en/next/handling-safe-area.html)
> 
> Fortunately, [expo support `react-native-safe-area-context` library.](https://docs.expo.io/versions/v35.0.0/sdk/safe-area-context/)
> 
> I will try this method within today and add commit if it works.
> 
> > @JongtaekChoi How is this going?

It doesn't work.

hyochan commented:
@JongtaekChoi Sad news.. Then what if changing your modal into screen component?

JongtaekChoi commented:
I'll try a little more.

JongtaekChoi commented:
@hyochan  I've made a change to the headers of UpdateProfile and Setting screen. 
I found that the Chat and SearchUser screens have the same issues and I fixed them. 
It may be a temporal solution and some similar issues may exist. 
But I think this pr is enough to long and these issues could be fixed at another pr or commit.

hyochan commented on src/components/screen/ChangePw/index.tsx:
```diff
@@ -0,0 +1,131 @@
+            <Ionicons name="md-close" size={24} color={theme.fontColor} />
+          </View>
+        </CloseButton>
+      </Header>
```
I think removing this part will prevent such side effects on the header. Try using header which is already implemented in `react-navigation` that will handle such cases.

JongtaekChoi commented:
@hyochan Thanks for your review. 
I made it by adding options to the ChangePw screen stack.
And I set `headerMode` of MainStackNavigator to 'screen' because of this [document](https://reactnavigation.org/docs/en/next/stack-navigator.html#header). 
![image](https://user-images.githubusercontent.com/17980230/71141850-d4bf6180-2258-11ea-973c-437c825d1282.png)
It makes the header attached to the screen. 
so the bug you mentioned at [here](https://github.com/dooboolab/hackatalk-mobile/pull/58#pullrequestreview-318349931) also be resolved by this `screen` option. 
![Dec-19-2019 12-25-46](https://user-images.githubusercontent.com/17980230/71142474-cffbad00-225a-11ea-93be-01f913f17819.gif)

So I request you to review my codes. 
Thanks again.

hyochan commented on src/components/navigation/AuthStackNavigator.tsx:
```diff
@@ -11,26 +11,28 @@ const Stack = createStackNavigator();
-        },
-        headerTintColor: theme.fontColor,
-      }}
-    >
```
@JongtaekChoi Why did you remove this? It will remove the dark mode behavior in the header as shown below?

![image](https://user-images.githubusercontent.com/27461460/71303040-a37c9800-23f6-11ea-8d72-e5ffe1e63d66.png)

JongtaekChoi replied:
@hyochan Oh it's my mistake. 
When I was finding the cause of error about the header option, I tested if the secreenOptions be removed then the error occurred. And I forgot to revert this.
